tests/shouldwork/Cores/Xilinx/Floating/TH.hs

```diff
@@ -0,0 +1,339 @@
+    , encodeFloat (2 ^ (digits - 1) - 1) (minExp - digits)
+    , 0
+    )
+    -- Round to nearest
+    --
+    -- For a datatype with 4 bits of precision, the significands align as:
+    -- 1000
+    --     1001
+    -- -------- +
+    -- 1001
+  , ( 2 ^ (digits - 1)
+    , encodeFloat (2 ^ (digits - 1) + 1) (-digits)
+    , 2 ^ (digits - 1) + 1
+    )
+    -- 1000
+    --     01111
+    -- --------- +
+    -- 1000
+  , ( 2 ^ (digits - 1)
+    , encodeFloat (2 ^ digits - 1) (-digits - 1)
+    , 2 ^ (digits - 1)
+    )
+    -- Ties to even
+    --
+    -- 1000
+    --     1000
+    -- -------- +
+    -- 1000
+  , ( 2 ^ (digits - 1)
+    , encodeFloat (2 ^ (digits - 1)) (-digits)
+    , 2 ^ (digits - 1)
+    )
+    -- Round to nearest
+    --
+    -- For a datatype with 4 bits of precision, the significands align as:
+    -- 1001
+    --     1001
+    -- -------- +
+    -- 1010
+  , ( 2 ^ (digits - 1) + 1
+    , encodeFloat (2 ^ (digits - 1) + 1) (-digits)
+    , 2 ^ (digits - 1) + 2
+    )
+    -- 1001
+    --     01111
+    -- --------- +
+    -- 1001
+  , ( 2 ^ (digits - 1) + 1
+    , encodeFloat (2 ^ digits - 1) (-digits - 1)
+    , 2 ^ (digits - 1) + 1
+    )
+    -- Ties to even
+    --
+    -- 1001
+    --     1000
+    -- -------- +
+    -- 1010
+  , ( 2 ^ (digits - 1) + 1
+    , encodeFloat (2 ^ (digits - 1)) (-digits)
+    , 2 ^ (digits - 1) + 2
+    )
+    -- Rounding at maximum exponent
+    --
+    -- 1111
+    --     1000
+    -- -------- +
+    -- infinity
+  , ( encodeFloat (2 ^ digits - 1) (maxExp - digits)
+    , encodeFloat (2 ^ (digits - 1)) (maxExp - 2*digits)
+    , 1/0
+    )
+    -- 1111
+    --     01111
+    -- --------- +
+    -- 1111
+  , ( encodeFloat (2 ^ digits - 1) (maxExp - digits)
+    , encodeFloat (2 ^ digits - 1) (maxExp - 2*digits - 1)
+    , encodeFloat (2 ^ digits - 1) (maxExp - digits)
+    )
+    -- Infinities
+  , (1/0, 1, 1/0)
+  , (-1/0, 1, -1/0)
+  , (1/0, -1/0, F.xilinxNaN)
+  ]
+  ++ cartesianProductTest model interesting interesting
+  ++ nanTest
+ where
+  digits = floatDigits (undefined :: Float)
+  (minExp, maxExp) = floatRange (undefined :: Float)
+  model (conditionFloat -> x) (conditionFloat -> y) = conditionFloat $ x + y
+
+cartesianProductTest
+  :: (a -> b -> c)
+  -> [a]
+  -> [b]
+  -> [(a,b,c)]
+cartesianProductTest f as bs =
+  map (\(a,b) -> (a, b, f a b)) $ cartesianProduct as bs
+
+cartesianProduct
+  :: [a]
+  -> [b]
+  -> [(a,b)]
+cartesianProduct as bs =
+  concatMap (\a -> map (\b -> (a,b)) bs) as
+
+interesting :: [Float]
+interesting =
+  [ -- Infinity
+    1/0
+    -- Minimum subnormal
+  , encodeFloat 1 (minExp - digits)
+    -- Maximum subnormal
+  , encodeFloat (2 ^ (digits - 1) - 1) (minExp - digits)
+    -- Minimum normal
+  , encodeFloat 1 (minExp - 1)
+    -- Maximum normal
+  , encodeFloat (2 ^ digits - 1) (maxExp - digits)
+    -- Epsilon (smallest positive x such that 1 + x is representable)
+  , encodeFloat 1 (1 - digits)
```

Comment:
Perhaps a dumb question, but why not use the names from https://hackage.haskell.org/package/ieee754-0.8.0/docs/Numeric-IEEE.html?

Reply:
Because they don't define anywhere near everything I want to use and I thought consistency was more desirable than being able to use a few short function names when they were available.

Reply:
I've gone back on my earlier decision; your comment made me realise I actually prefer using the `Numeric.IEEE` names where possible. This is more readable.